src/Language/PureScript/TypeChecker/Deriving.hs

```diff
@@ -0,0 +1,150 @@
+      mkSigRec =  App (Constructor $ findName "SigRecord" ctors) . ArrayLiteral
+      genericModule = case findName "Generic" tcs of (Qualified mn _) -> mn
+
+      mkProdClause (ctorName, (_,_,typ,_)) = ObjectLiteral [("sigConstructor",StringLiteral (runProperName . unQualify $ ctorName)),("sigValues", ArrayLiteral . map mkProductSignature . argTypes $ typ)]
+
+      mkProductSignature (ObjectType rec) =
+          lam "q" . mkSigRec .
+              map (\(str,typ) -> ObjectLiteral [("recLabel", StringLiteral str), ("recValue", mkProductSignature typ)])
+              $ decomposeRec rec
+      mkProductSignature typ = lam "q" $ App (mkVarMn genericModule "toSignature")
+                               (TypedValue False (mkVarMn genericModule "anyProxy") (TypeApp (TypeConstructor $ findName "Proxy" envtyps) typ))
+
+  return . lam "x" . mkSigProd . map mkProdClause . constructorsForType t $ ctors
+
+mkFromSpineFunction :: (Functor m, MonadState CheckState m) => Qualified ProperName ->  m Expr
+mkFromSpineFunction t = do
+  ctors <- M.toList . dataConstructors <$> getEnv
+  tcs <- M.toList . typeClasses <$> getEnv
+  let
+      mkJust    = App (Constructor $ findName "Just" ctors)
+      mkNothing = Constructor $ findName "Nothing" ctors
+      genericModule = case findName "Generic" tcs of (Qualified mn _) -> mn
+
+      mkAlternative (ctor@(Qualified _ (ProperName ctorName)), (_,_,typ,idents)) =
+          CaseAlternative [ConstructorBinder (findName "SProd" ctors) [StringBinder ctorName, ArrayBinder (map VarBinder idents)]]
+                          . Right $ liftApplicative (mkJust $ Constructor ctor) (zipWith fromSpineFun (map (Var . (Qualified Nothing))idents) (argTypes typ))
+
+      fromSpineFun e (ObjectType rec) = App (lamCase "r" [mkRecCase (decomposeRec rec), CaseAlternative [NullBinder] (Right mkNothing)]) (App e (mkPrelVar "unit"))
+
+      fromSpineFun e _ = App (mkVarMn genericModule "fromSpine") (App e (mkPrelVar "unit"))
+
+      mkRecCase rs = CaseAlternative [ConstructorBinder (findName "SRecord" ctors) [ArrayBinder (map (VarBinder . Ident . fst) rs)]] . Right $ liftApplicative (mkRecFun rs) (map (\(x,y) -> fromSpineFun (Accessor "recValue" (mkVar x)) y) rs)
+
+      mkRecFun :: [(String,Type)] -> Expr
+      mkRecFun xs = mkJust $ foldr (\s e -> lam s e) recLiteral (map fst xs)
+         where recLiteral = ObjectLiteral $ map (\(s,_) -> (s,mkVar s)) xs
+
+  return . lamCase "x" $ map mkAlternative (constructorsForType t ctors) ++ [CaseAlternative [NullBinder] (Right mkNothing)]
+
+-- Helpers
+
+pattern ObjectType rec = TypeApp (TypeConstructor (Qualified (Just (ModuleName [ProperName "Prim"])) (ProperName "Object"))) rec
+
+lam :: String -> Expr -> Expr
+lam s = Abs (Left (Ident s))
+
+lamCase :: String -> [CaseAlternative] -> Expr
+lamCase s = lam s . Case [mkVar s]
+
+liftApplicative :: Expr -> [Expr] -> Expr
+liftApplicative = foldl' (\x e -> App (App (mkPrelVar "ap") x) e)

```

Comment:
Maybe use `apply` here to avoid an indirection via `Monad` in the generated code?